### PR TITLE
Augment Coordinates for periodic Boundary conditions

### PR DESCRIPTION
**Pull request by ayushsuhane**
Changes made in this Pull Request:
- Cython implementation of adding duplicate particles for periodic boundary conditions
- class Periodic_cKDTree  similar to PeriodicKDTree but a wrapper around more fast scipy.spatial.cKDTree.
 
PR Checklist
------------
 - [ ] Tests?
 - [ ] Docs?
 - [ ] CHANGELOG updated?
 - [ ] Issue raised/referenced?

### CODE REVIEW

**Comment by ayushsuhane** [comment]
Since replacement of Bio.cKDTree with scipy.spatial.cKDTree is beneficial, is this the correct way forward. The  class Periodic_cKDTree deals with both periodic and non-periodic boundary conditons. I have also defined a `search_query` function which can directly be used in guess_bonds in `MDAnalysis.topology.guessers.py`. With capped_distance and Periodic_cKDTree, the `selection.py` can be rewritten in a way such that it would be easy to include the grid search algorithm in it as well.

File: package/MDAnalysis/lib/c_distances.pyx [hunk 1]
```diff
@@ -315,3 +317,290 @@ def contact_matrix_pbc(coord, sparse_contacts, box, cutoff):
...
+def augment(float[:, ::1] coordinates, float[:,::1] box, float[:,::1] reciprocal, float r):
+    """Calculate augmented coordinate set
+    
+    Parameters
+    ----------
+    coordinates : array
+      Input coordinate array to generate duplicate images
+    box : array
+      Dimensions of the box of shape (3, 3)
+    reciprocal : array
```
Comment (richardjgowers) [inline package/MDAnalysis/lib/c_distances.pyx]: is this the reciprocal of `box`?  if so, we can just calculate it in the function

File: package/MDAnalysis/lib/c_distances.pyx [hunk 2]
```diff
@@ -315,3 +317,290 @@ def contact_matrix_pbc(coord, sparse_contacts, box, cutoff):
...
+    cdef int i, j, p, N
+    cdef float shiftX[3]
+    cdef float shiftY[3]
+    cdef float shiftZ[3]
+    cdef float coord[3]
+    cdef float end[3]
+    cdef float other[3]
+    
+    cdef int dim
+    dim = coordinates.shape[1]
```
Comment (richardjgowers) [inline package/MDAnalysis/lib/c_distances.pyx]: we can hardcode dim to 3, it might sometimes help the compiler

File: package/MDAnalysis/lib/c_distances.pyx [hunk 3]
```diff
@@ -315,3 +317,290 @@ def contact_matrix_pbc(coord, sparse_contacts, box, cutoff):
...
+                    p += 1
+                    
+                    
+                elif hi_z:
+                    for j in range(3):
+                        output[p, j] = coord[j] - shiftX[j] + shiftY[j] - shiftZ[j]
+                    indices[p] = i
+                    p += 1
+                    
+
```
Comment (richardjgowers) [inline package/MDAnalysis/lib/c_distances.pyx]: if you can fix up the formatting it would be easier to read, there's 3 blank lines here

File: package/MDAnalysis/lib/c_distances.pyx [hunk 4]
```diff
@@ -315,3 +317,290 @@ def contact_matrix_pbc(coord, sparse_contacts, box, cutoff):
...
+@cython.boundscheck(False)
+@cython.wraparound(False)
+cdef float _dot(float  *a, float *b):
+    """Return dot product of two sequences in range."""
+    cdef ssize_t n
+    cdef float sum1
+    cdef ssize_t dim
+    
+    dim=3
```
Comment (richardjgowers) [inline package/MDAnalysis/lib/c_distances.pyx]: hardcode dim to 3

File: package/MDAnalysis/lib/c_distances.pyx [hunk 5]
```diff
@@ -32,6 +32,8 @@ cimport cython
 import numpy
 cimport numpy
 
+from libc.math cimport sqrt
```
Comment (richardjgowers) [inline package/MDAnalysis/lib/c_distances.pyx]: unused

File: package/MDAnalysis/lib/c_distances.pyx [hunk 6]
```diff
@@ -315,3 +317,290 @@ def contact_matrix_pbc(coord, sparse_contacts, box, cutoff):
...
+        sum1 += a[n] * b[n]
+    return sum1
+
+
+@cython.boundscheck(False)
+@cython.wraparound(False)
+def undo_augment(int[:] results, int[:] translation, int nreal):
+    """Translate augmented indices back to originals
+    
+    Note: modifies results in place!
```
[the remark this answers is not in the document]
Comment (ayushsuhane) [reply package/MDAnalysis/lib/c_distances.pyx]: Actually, I thought for complex selections it might be useful to keep all the particles (original + images). But it looks like its better to modify the results, otherwise it will lead to multiple computations of similar points with increase in number of selections(operations). 
Meanwhile, I will also try to think of a  case, where it might fail.

**Comment by ayushsuhane** [comment]
I was working on the tests, but got stuck in few problems and would like to share them here:

1) As you might see in the augment function, the shape of output array is (N, 3). If I supply only a single coordinate close to the face, and/or edge, it will not be able to calculate all the periodic images. Possible solutions are to increase the size of output array by preferably 6 times, but it will effect the performance for large array of particles. Another solution could be [dynamic allocation](https://github.com/cython/cython/wiki/DynamicMemoryAllocation) of memory for output and indices. 

2) This function has another parameter cutoff, which is required while creating the tree. What could be a sensible default value for it? 

3) Instead of augmenting the coordinates, other option could be to cythonize the find_images function which should also improve the performance i.e. directly returning the minimum distance through the function.

**Comment by jbarnoud** [comment]
> As you might see in the augment function, the shape of output array is (N, 3). If I supply only a single coordinate which lies close to the face and/or edge, it will not be able to calculate all the periodic images. Possible solutions are to increase the size of output array by preferably 6 times, but it will effect the performance for large array of particles. Another solution could be dynamic allocation of memory for output and indices.

I do not understand the problem. Could you elaborate?

**Comment by ayushsuhane** [comment]
For instance, if the augment function is called as `augment(a, box, cutoff)` where `a = np.array([[1,1,1]], dtype = np.float32)`, `box = np.array([10, 10, 10, 90, 90, 90], dtype=np.float32)` and radius > 2. It should ideally create 6 images points, but since the output array is allocated as ` cdef float[:, :] output = np.zeros((N, 3), dtype=np.float32)`, it cannot store all the images and the function, in its current form will return only 1 image.

**Comment by ayushsuhane** [comment]
Yes, Exactly. For larger number of atoms, it can be assumed that few particles will be close to the boundary and therefore, it works, but the current form will fail for few number of particles. Either we can put a condition that `output` should be minimum 100 or 1000, which wouldn't have an influence on the performance and will also yield correct number of images for evenly distributed data. But if all the edge cases needs to be considered, then we need to approach it from a different strategy.

**Comment by jbarnoud** [comment]
At which point in the code do you know for sure what the size of `output` should be?

**Comment by ayushsuhane** [comment]
Only at the end. That is why it is allocated a size of `(N,3)` but the size of returned array is `(p,3)` where, it is assumed that p<N

**Comment by jbarnoud** [comment]
Can you really assume p < N? If you can have p = 6 for N = 1, then cannot you have p = 6 * N? What happens if all your points are at the border of the box?

**Comment by richardjgowers** [comment]
Yeah this is something I forgot to do, I think I have a solution....

**Comment by richardjgowers** [comment]
@ayushsuhane ok I think we can just use a cpp vector of floats and ints

```python
from libcpp.vector cimport vector

import numpy as np


def thing():
    cdef vector[float] output
    cdef vector[int] indices
    
    # each time we add an augmented coordinate
    for j in range(3):
        output.push_back(coord[j] + shiftX[j] + shiftY[j])
    indices.push_back(i)
    
    # at the end to return results
    n = indices.size()

    return np.asarray(output).reshape(n, 3), np.asarray(indices)

File: package/MDAnalysis/lib/c_distances.pyx [hunk 7]
```diff
@@ -315,3 +317,307 @@ def contact_matrix_pbc(coord, sparse_contacts, box, cutoff):
...
+
+@cython.boundscheck(False)
+@cython.wraparound(False)
+cdef float[:] _cross(float * a, float * b):
+    """
+    Calculates the cross product between vectors
+    given by pointers a and b
+    """
+    cdef ssize_t n
+    cdef float[:] result = numpy.zeros((3,), dtype=numpy.float32)
```
Comment (richardjgowers) [inline package/MDAnalysis/lib/c_distances.pyx]: in general you don't want any numpy calls inside a cdef function, this should all just be pure C.  We only want numpy calls at the start of a function (to prepare things) and at the end (to correctly format things for the user)

**Comment by ayushsuhane** [comment]
`np.unique(array, axis=0)`  fails the tests. Should I change it, or add upgrade numpy during the installation in Tarvis.

File: package/MDAnalysis/lib/pkdtree.py [hunk 8]
```diff
@@ -230,3 +232,207 @@ def get_indices(self):
...
+                                             self.mapping,
+                                             len(self.coords))
+        else:
+            wrapped_centers = np.asarray(centers)
+            indices = list(self.ckdt.query_ball_point(wrapped_centers,
+                                                      radius))
+            self._indices = np.array(list(
+                                     itertools.chain.from_iterable(indices)),
+                                     dtype=np.int32)
+        self._indices = np.unique(self._indices)
```
Comment (richardjgowers) [inline package/MDAnalysis/lib/pkdtree.py]: there's a faster unique in `lib._cutil` that we can use

File: package/MDAnalysis/lib/pkdtree.py [hunk 9]
```diff
@@ -230,3 +232,207 @@ def get_indices(self):
...
+            if len(pairs) > 1:
+                pairs[:, 0] = undo_augment(pairs[:, 0], self.mapping,
+                                           len(self.coords))
+                pairs[:, 1] = undo_augment(pairs[:, 1], self.mapping,
+                                           len(self.coords))
+        else:
+            pairs = np.array(list(self.ckdt.query_pairs(radius)),
+                             dtype=np.int32)
+        if pairs.size > 0:
+            pairs = np.unique(np.sort(pairs), axis=0)
```
Comment (richardjgowers) [inline package/MDAnalysis/lib/pkdtree.py]: use the cutil.unique one here too

File: package/MDAnalysis/lib/c_distances.pyx [hunk 10]
```diff
@@ -20,7 +20,6 @@
 # J. Comput. Chem. 32 (2011), 2319--2327, doi:10.1002/jcc.21787
 #
 #
```
Comment (richardjgowers) [inline package/MDAnalysis/lib/c_distances.pyx]: can you revert the changes to this file to keep history tidy?

File: package/MDAnalysis/lib/pkdtree.py [hunk 11]
```diff
@@ -230,3 +232,207 @@ def get_indices(self):
           neighbors for the last query points and search radius
         """
         return self._indices
+
+
+class Periodic_cKDTree(object):
```
Comment (richardjgowers) [inline package/MDAnalysis/lib/pkdtree.py]: We need to replace the existing pkdtree rather than just add another option here

File: package/MDAnalysis/lib/c_distances.pyx (hunk 10 above)
Comment (ayushsuhane) [reply package/MDAnalysis/lib/c_distances.pyx]: Just to make sure, are you asking to commit a revert or to make a separate branch and push force to the current branch, but I guess you have to rebase it afterwards to not include any of the current history.

**Comment by richardjgowers** [comment]
@ayushsuhane yeah just a PR with only the augment functions and tests, but don't lose the progress you've made on KDTree, it's not too far from being finished either

**Comment by ayushsuhane** [comment]
@richardjgowers  Do we need anything else here? 

I will open another PR for replacing Bio.KDTree .

File: package/MDAnalysis/lib/_augment.pyx [hunk 12]
```diff
@@ -0,0 +1,307 @@
...
+
+    Note
+    ----
+    Modifies the results array
+
+    """
+    cdef int N
+    N = results.shape[0]
+
+    for i in range(N):
```
Comment (richardjgowers) [inline package/MDAnalysis/lib/_augment.pyx]: `i` isn't cdef'd here

File: package/MDAnalysis/lib/_augment.pyx [hunk 13]
```diff
@@ -0,0 +1,307 @@
...
+
+    Returns
+    -------
+    output : array
+      coordinates of duplicates generated due to periodic boundary conditions
+    indices : array
+      original indices of the augmented coordinates
+    """
+    cdef bint lo_x, hi_x, lo_y, hi_y, lo_z, hi_z
+    cdef int i, j, p, N
```
Comment (richardjgowers) [inline package/MDAnalysis/lib/_augment.pyx]: `p` is unused

File: package/MDAnalysis/lib/_augment.pyx [hunk 14]
```diff
@@ -0,0 +1,307 @@
...
+#
+# N. Michaud-Agrawal, E. J. Denning, T. B. Woolf, and O. Beckstein.
+# MDAnalysis: A Toolkit for the Analysis of Molecular Dynamics Simulations.
+# J. Comput. Chem. 32 (2011), 2319--2327, doi:10.1002/jcc.21787
+#
+#
+
+import cython
+import numpy
+cimport numpy
```
Comment (richardjgowers) [inline package/MDAnalysis/lib/_augment.pyx]: We're never using the cimport of numpy here.  We'd be using it if we did something like `cdef numpy.ndarray[float] arr1`, the numpy calls on the right hand side use the `import numpy` import

File: package/MDAnalysis/lib/_augment.pyx [hunk 15]
```diff
@@ -0,0 +1,307 @@
...
+# Python in Science Conference, pages 102-109, Austin, TX, 2016. SciPy.
+#
+# N. Michaud-Agrawal, E. J. Denning, T. B. Woolf, and O. Beckstein.
+# MDAnalysis: A Toolkit for the Analysis of Molecular Dynamics Simulations.
+# J. Comput. Chem. 32 (2011), 2319--2327, doi:10.1002/jcc.21787
+#
+#
+
+import cython
+import numpy
```
Comment (richardjgowers) [inline package/MDAnalysis/lib/_augment.pyx]: `as np`, then change the calls to `np.thing`

File: package/MDAnalysis/lib/_augment.pyx [hunk 16]
```diff
@@ -0,0 +1,307 @@
...
+@cython.boundscheck(False)
+@cython.wraparound(False)
+def augment(float[:, ::1] coordinates, float[:, ::1] dm, float r):
+    """Calculate augmented coordinate set
+
+    Parameters
+    ----------
+    coordinates : array
+      Input coordinate array to generate duplicate images
+    dm : array
```
Comment (richardjgowers) [inline package/MDAnalysis/lib/_augment.pyx]: most other functions take box in the `[lx, ly, lz,, 90 90 90]` form, so use that form here.  Then do the conversion to triclinic vectors inside the function.  You can keep the `dm` variable, just have it created inside the function

File: package/CHANGELOG [hunk 17]
```diff
@@ -20,6 +20,7 @@ The rules for this file:
 
 Enhancements
 
+  * Added augment functionality to handle periodic boundary conditions (PR #1977)
```
Comment (richardjgowers) [inline package/CHANGELOG]: This doesn't really describe what the function does

File: package/MDAnalysis/lib/_augment.pyx [hunk 18]
```diff
@@ -0,0 +1,307 @@
...
+from libcpp.vector cimport vector
+from libc.math cimport sqrt
+
+__all__ = ['augment', 'undo_augment']
+
+@cython.boundscheck(False)
+@cython.wraparound(False)
+def augment(float[:, ::1] coordinates, float[:, ::1] dm, float r):
+    """Calculate augmented coordinate set
+
```
Comment (richardjgowers) [inline package/MDAnalysis/lib/_augment.pyx]: Add: "Note, does not return coordinates from the initial array, to merge these use `np.concatenate(a, b)`" (or is it `np.concat`?)

File: package/MDAnalysis/lib/_augment.pyx [hunk 19]
```diff
@@ -0,0 +1,307 @@
...
+    r : float
+      thickness of cutoff region for duplicate image generation
+
+    Returns
+    -------
+    output : array
+      coordinates of duplicates generated due to periodic boundary conditions
+    indices : array
+      original indices of the augmented coordinates
+    """
```
Comment (richardjgowers) [inline package/MDAnalysis/lib/_augment.pyx]: `.. versionadded:: 0.19.0`

File: package/MDAnalysis/lib/_augment.pyx [hunk 20]
```diff
@@ -0,0 +1,307 @@
...
+
+    Returns
+    -------
+    results : ndarray of ints
+
+    Note
+    ----
+    Modifies the results array
+
+    """
```
Comment (richardjgowers) [inline package/MDAnalysis/lib/_augment.pyx]: `.. versionadded:: 0.19.0`

File: package/MDAnalysis/lib/_augment.pyx [hunk 21]
```diff
@@ -0,0 +1,307 @@
...
+      than this need to be translated to their real counterpart
+
+    Returns
+    -------
+    results : ndarray of ints
+
+    Note
+    ----
+    Modifies the results array
+
```
Comment (richardjgowers) [inline package/MDAnalysis/lib/_augment.pyx]: `.. seealso:: augment_coordinates`

File: package/MDAnalysis/lib/_augment.pyx [hunk 22]
```diff
@@ -0,0 +1,307 @@
...
+      number of real coordinates, ie values in results equal or larger
+      than this need to be translated to their real counterpart
+
+    Returns
+    -------
+    results : ndarray of ints
+
+    Note
+    ----
+    Modifies the results array
```
Comment (richardjgowers) [inline package/MDAnalysis/lib/_augment.pyx]: in place

File: package/MDAnalysis/lib/_augment.pyx [hunk 23]
```diff
@@ -0,0 +1,307 @@
...
+      indices of coordinates, including "augmented" indices
+    translation : ndarray of ints
+      original indices of augmented coordinates
+    nreal : int
+      number of real coordinates, ie values in results equal or larger
+      than this need to be translated to their real counterpart
+
+    Returns
+    -------
+    results : ndarray of ints
```
Comment (richardjgowers) [inline package/MDAnalysis/lib/_augment.pyx]: indices which have been translated to refer to the original indices

File: package/MDAnalysis/lib/_augment.pyx [hunk 24]
```diff
@@ -0,0 +1,307 @@
...
+      Vectors `[[a], [b], [c]]` are box vectors
+    r : float
+      thickness of cutoff region for duplicate image generation
+
+    Returns
+    -------
+    output : array
+      coordinates of duplicates generated due to periodic boundary conditions
+    indices : array
+      original indices of the augmented coordinates
```
Comment (richardjgowers) [inline package/MDAnalysis/lib/_augment.pyx]: `.. seealso:: undo_augment`

File: package/MDAnalysis/lib/_augment.pyx [hunk 25]
```diff
@@ -0,0 +1,307 @@
...
+cimport numpy
+
+from libcpp.vector cimport vector
+from libc.math cimport sqrt
+
+__all__ = ['augment', 'undo_augment']
+
+@cython.boundscheck(False)
+@cython.wraparound(False)
+def augment(float[:, ::1] coordinates, float[:, ::1] dm, float r):
```
Comment (richardjgowers) [inline package/MDAnalysis/lib/_augment.pyx]: rename this to `augment_coordinates` so it's clear it works on coordinates

File: package/MDAnalysis/lib/_augment.pyx (hunk 18 above)
Comment (richardjgowers) [inline package/MDAnalysis/lib/_augment.pyx]: Add a description of what this function does.  Something like "calculates which particles are within `r` of a box boundary and creates duplicate images of these on the opposite side of the box"

File: package/MDAnalysis/lib/_augment.pyx [hunk 26]
```diff
@@ -0,0 +1,307 @@
...
+    """Calculate augmented coordinate set
+
+    Parameters
+    ----------
+    coordinates : array
+      Input coordinate array to generate duplicate images
```
Comment (richardjgowers) [inline package/MDAnalysis/lib/_augment.pyx]: These must all be within the primary unit cell for the algorithm to work

File: package/MDAnalysis/lib/_augment.pyx (hunk 18 above)
Comment (ayushsuhane) [reply package/MDAnalysis/lib/_augment.pyx]: Do you think it should be added within the function itself since undo_augment already assumes that  results are after the merge? Or will there be any benefit in keeping it in the same way? One could, in principle, provide 0 as nreal, but I am not sure how will that be useful?

File: testsuite/MDAnalysisTests/lib/test_augment.py [hunk 27]
```diff
@@ -0,0 +1,100 @@
...
+           [1.1, 1.1, 1.1]),
+          ([1.1, 0.9, 0.1], [0.1, -0.1, 0.1], [0.1, 0.9, 1.1],
+           [0.1, -0.1, 1.1], [1.1, -0.1, 0.1], [1.1, 0.9, 1.1],
+           [1.1, -0.1, 1.1]),
+          ([1.1, 0.9, 0.1], [0.1, -0.1, 0.1], [0.1, 0.9, 1.1],
+           [0.1, -0.1, 1.1], [1.1, -0.1, 0.1], [1.1, 0.9, 1.1],
+           [1.1, -0.1, 1.1]),
+          ([1.1, 0.5, 0.5], [0.5, -0.1, 0.5]))
+
+radius = 1.5
```
Comment (richardjgowers) [inline testsuite/MDAnalysisTests/lib/test_augment.py]: this should be inside the test function

File: testsuite/MDAnalysisTests/lib/test_augment.py [hunk 28]
```diff
@@ -0,0 +1,100 @@
...
+
+
+@pytest.mark.parametrize('b, qres', product(boxes, zip(queries, images)))
+def test_augment(b, qres):
+    b = np.array(b, dtype=np.float32)
```
Comment (richardjgowers) [inline testsuite/MDAnalysisTests/lib/test_augment.py]: just store b as a numpy array?

File: testsuite/MDAnalysisTests/lib/test_augment.py [hunk 29]
```diff
@@ -0,0 +1,100 @@
...
+
+from MDAnalysis.lib._augment import augment_coordinates, undo_augment
+from MDAnalysis.lib.distances import apply_PBC, transform_StoR
+
+boxes = ([10, 10, 10, 90, 90, 90],  # ortho
+         [10, 10, 10, 45, 60, 90])  # tri_box
+
+# Find images for several query points,
+# here in fractional coordinates using augment
+queries = ([0.1, 0.5, 0.5],  # box face
```
Comment (richardjgowers) [inline testsuite/MDAnalysisTests/lib/test_augment.py]: can you reformat this so the queries and images are side by side, it's hard to read this way..

```python
qres = [
    ([[0.1, 0.5, 0.5]], [[1.1, 0.5 0.5]]),  # box face
    etc
]

```

File: testsuite/MDAnalysisTests/lib/test_augment.py [hunk 30]
```diff
@@ -0,0 +1,100 @@
...
+           [1.1, -0.1, 1.1]),
+          ([1.1, 0.9, 0.1], [0.1, -0.1, 0.1], [0.1, 0.9, 1.1],
+           [0.1, -0.1, 1.1], [1.1, -0.1, 0.1], [1.1, 0.9, 1.1],
+           [1.1, -0.1, 1.1]),
+          ([1.1, 0.5, 0.5], [0.5, -0.1, 0.5]))
+
+radius = 1.5
+
+
+@pytest.mark.parametrize('b, qres', product(boxes, zip(queries, images)))
```
Comment (richardjgowers) [inline testsuite/MDAnalysisTests/lib/test_augment.py]: don't use `product`, just have two `mark.parametrize`, one for `b` one for `qres`

File: package/setup.py [hunk 31]
```diff
@@ -353,6 +353,13 @@ def extensions(config):
                          include_dirs=include_dirs + ['MDAnalysis/lib/include'],
                          define_macros=define_macros,
                          extra_compile_args=cpp_extra_compile_args)
+    aug = MDAExtension('MDAnalysis.lib._augment',
+                         sources=['MDAnalysis/lib/_augment' + source_suffix],
+                         language='c++',
+                         libraries=mathlib,
```
Comment (richardjgowers) [inline package/setup.py]: can you check if we need mathlib?  I thought we needed it for sqrt

File: package/MDAnalysis/lib/_augment.pyx [hunk 32]
```diff
@@ -0,0 +1,353 @@
...
+    cdef bint lo_x, hi_x, lo_y, hi_y, lo_z, hi_z
+    cdef int i, j, N
+    cdef float norm
+    cdef float shiftX[3]
+    cdef float shiftY[3]
+    cdef float shiftZ[3]
+    cdef float coord[3]
+    cdef float end[3]
+    cdef float other[3]
+    cdef float[:, ::1] dm = np.zeros((3, 3), dtype=np.float32)
```
Comment (richardjgowers) [inline package/MDAnalysis/lib/_augment.pyx]: `dm` and `reciprocal` don't need to be numpy arrays, can we just use C floats?

**Comment by ayushsuhane** [comment]
Should the return datatype for indices be changed to int64 specifically? Or the current way of typecasting before calling unique_int_1d is workable.

**Comment by richardjgowers** [comment]
@ayushsuhane are you talking about how `unique_int` takes int64s but I've used int32s here?  Might be good to unify that tbh

**Comment by ayushsuhane** [comment]
Should it handle both or everything should be converted to int64?

**Comment by ayushsuhane** [comment]
Meanwhile, I just changed it to np.int64 , if required to handle both of the int types I came across  `ctypedef fused` which can be used. 

@richardjgowers @jbarnoud  can you review it. Once it is merged then we can focus on the KDTree, since it also require augment function.

File: package/doc/sphinx/source/documentation_pages/lib/augment.rst [hunk 33]
```diff
@@ -0,0 +1,2 @@
+.. automodule:: MDAnalysis.lib.augment
```
Comment (richardjgowers) [inline package/doc/sphinx/source/documentation_pages/lib/augment.rst]: the doc build failed because you refer to `lib.augment` here but you named the module `lib._augment` elsewhere

**Comment by richardjgowers** [comment]
@ayushsuhane WRT `dot` and `cross`.  If you could put them into `_cutil` and make a header file (`pxd`) that would also be good.  If you could use fused types for them (float/double) that would be even better.  If it ends up being difficult we can do it in a later PR, I'd like to get this merged

**Comment by ayushsuhane** [comment]
@richardjgowers were you referring to similar structure for the docs or did I get you wrong?

**Comment by richardjgowers** [comment]
@ayushsuhane yep that was right, thanks!
